discovery_engine_core/core/src/engine.rs

```diff
@@ -92,60 +97,118 @@ where
-    where
-        R: Send + Sync,
-    {
-        Selection::new(BetaSampler)
-            .select(self.stacks.values_mut().collect(), max_documents)
+    #[allow(clippy::future_not_send)]
+    pub async fn get_feed_documents(&self, max_documents: usize) -> Result<Vec<Document>, Error> {
+        Selection::new(BetaSampler, self.stacks.values())
+            .select(max_documents)
+            .await
             .map_err(|e| e.into())
     }
 
     /// Process the feedback about the user spending some time on a document.
-    pub fn time_logged(&mut self, time_logged: &TimeSpent) -> Result<(), Error> {
-        self.ranker.time_logged(time_logged)?;
+    #[allow(clippy::future_not_send)]
+    pub async fn time_spent(&mut self, time_spent: &TimeSpent) -> Result<(), Error> {
+        self.ranker.time_spent(time_spent)?;
 
-        rank_stacks(self.stacks.values_mut(), &self.ranker)
+        rank_stacks(self.stacks.values(), &self.ranker).await
     }
 
     /// Process the feedback about the user reacting to a document.
-    pub fn user_reacted(&mut self, reacted: &UserReacted) -> Result<(), Error> {
-        let stack = self
-            .stacks
-            .get_mut(&reacted.stack_id)
-            .ok_or(Error::InvalidStackId(reacted.stack_id))?;
-
-        stack.update_relevance(reacted.reaction);
+    #[allow(clippy::future_not_send)]
+    pub async fn user_reacted(&mut self, reacted: &UserReacted) -> Result<(), Error> {
+        self.stacks
+            .get(&reacted.stack_id)
+            .ok_or(Error::InvalidStackId(reacted.stack_id))?
+            .write()
+            .await
+            .update_relevance(reacted.reaction);
 
         self.ranker.user_reacted(reacted)?;
 
-        rank_stacks(self.stacks.values_mut(), &self.ranker)
+        rank_stacks(self.stacks.values(), &self.ranker).await
+    }
+
+    /// Updates the stacks with data related to the top key phrases of the current data.
+    #[allow(dead_code)]
+    #[allow(clippy::future_not_send)]
+    async fn update_stacks(&mut self, top: usize) -> Result<(), Error> {
+        let key_phrases = self.ranker.select_top_key_phrases(top);
+        let key_phrases = &key_phrases;
+        let ranker = &self.ranker;
+
+        self.stacks
+            .values()
+            .map(|stack| async move {
+                stack
+                    .read()
+                    .await
+                    .ops
+                    .new_items(key_phrases, ranker)
+                    .await
+                    .map(|documents| (stack, documents))
+            })
+            .collect::<FuturesUnordered<_>>()
+            .try_for_each(|(stack, documents)| async move {
+                stack
+                    .write()
+                    .await
+                    .update(&documents, ranker)
+                    .map_err(GenericError::from)
+            })
+            .await
+            .map_err(Into::into)
```

Comment:
It might not matter for us (does it?) but this code can have race conditions.

Mainly two tasks might calculated a update for the same stack at the same time
and then write it. Whoever writes first will be overwritten by whoever writes second.

Reply:
within the method all stacks should be distinct (as they come from the hashmap) and they are updated individually. from the outside only a single task should be able to call the method at a time (as the receiver is `&mut self`). so i'm not sure if i can see where the race condition could come from?

in any case, i agree that we have to very carefully review the async code to avoid race conditions. for now i was told that we shouldn't spend too much time on this as it is more important to have a first running version.

Reply:
With "it might not  matter for us" I meant exactly things like "the way we use this this situation can't happen".

Reply:
"it might not matter for us" and  "the way we use this this situation can't happen" have two really different meaning.
The method take a `&mut` so this code cannot have a race condition.